misc/python/mzbuild.py

```diff
@@ -0,0 +1,359 @@
+
+    def depends(self, root: Path, path: Path) -> List[bytes]:
+        pass
+
+
+class CargoBuild(PreImage):
+    def __init__(self, config: Dict[str, Any]):
+        self.bin = config.pop("bin", None)
+        self.strip = config.pop("strip", True)
+        if self.bin is None:
+            raise ValueError("mzbuild config is missing pre-build target")
+
+    def run(self, root: Path, path: Path) -> None:
+        super().run(root, path)
+        runv([xcargo(root), "build", "--release", "--bin", self.bin], cwd=root)
+        shutil.copy(str(xcargo_target_dir(root) / "release" / self.bin),
+                    str(path))
+        if self.strip:
+            # NOTE(benesch): the debug information is large enough that it slows
+            # down CI, since we're packaging these binaries up into Docker
+            # images and shipping them around. A bit unfortunate, since it'd be
+            # nice to have useful backtraces if the binary crashes.
+            runv([xstrip(root), path / self.bin])
+
+    def depends(self, root: Path, path: Path) -> List[bytes]:
+        # TODO(benesch): this should be much smarter about computing the Rust
+        # files that actually contribute to this binary target.
+        return git_ls_files(root, "src/**", "Cargo.toml", "Cargo.lock")
+
+
+# TODO(benesch): make this less hardcoded and custom.
+class CargoTest(PreImage):
+    def __init__(self, config: Dict[str, Any]):
+        pass
+
+    def run(self, root: Path, path: Path) -> None:
+        super().run(root, path)
+        CargoBuild({"bin": "testdrive"}).run(root, path)
+
+        # NOTE(benesch): The two invocations of `cargo test --no-run` here
+        # deserve some explanation. The first invocation prints error messages
+        # to stdout in a human readable form. If that succeeds, the second
+        # invocation instructs Cargo to dump the locations of the test binaries
+        # it built in a machine readable form. Without the first invocation, the
+        # error messages would also be sent to the output file in JSON, and the
+        # user would only see a vague "could not compile <package>" error.
+        args = [xcargo(root), "test", "--locked",
+                "--no-run"]  # type: List[Union[str,Path]]
+        runv(args)
+
+        tests = []
+        for message in capture(args + ["--message-format=json"],
+                               universal_newlines=True).split("\n"):
+            if message.strip() == "":
+                continue
+            message = json.loads(message)
+            if message.get("profile", {}).get("test", False):
+                crate_name = message["package_id"].split()[0]
+                target_kind = "".join(message["target"]["kind"])
+                slug = crate_name + "." + target_kind
+                if target_kind != "lib":
+                    slug += "." + message["target"]["name"]
+                crate_path_match = re.search("\(path\+file://(.*)\)",
```

Comment:
this isn't buggy as written, but it's a good practice to use  raw strings (`r".."`) in regexes to make sure that backslash escapes never get pre-processed before they get to the regex engine.